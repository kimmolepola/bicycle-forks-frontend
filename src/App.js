import React, { useState, useEffect } from 'react';
import { ThemeProvider, makeStyles } from '@material-ui/core/styles';
import {
  CssBaseline, Hidden, Typography, Link, Snackbar, Slide,
} from '@material-ui/core';
import { Alert } from '@material-ui/lab';
import {
  useMutation, useQuery, gql,
} from '@apollo/client';

import Navigator from './components/Navigator';
import Map from './components/Map';
import Header from './components/Header';
import Theme from './Theme';
import Points from './components/Points';

const drawerWidth = 256;

const Copyright = () => (
  <Typography variant="body2" color="textSecondary" align="center">
    {'Copyright Â© '}
    <Link color="inherit" href="https://material-ui.com/">
      Your Website
    </Link>{' '}
    {new Date().getFullYear()}
    {'.'}
  </Typography>
);

const useStyles = makeStyles({
  root: {
    display: 'flex',
    minHeight: '100vh',
  },
  drawer: {
    [Theme.breakpoints.up('sm')]: {
      width: drawerWidth,
      flexShrink: 0,
    },
  },
  app: {
    flex: 1,
    display: 'flex',
    flexDirection: 'column',
  },
  main: {
    display: 'flex',
    flex: 1,
    paddingTop: Theme.spacing(6),
    paddingLeft: Theme.spacing(4),
    paddingRight: Theme.spacing(4),
    background: '#eaeff1',
  },
  footer: {
    padding: Theme.spacing(2),
    background: '#eaeff1',
  },
  header: {
    flex: 1,
  },
});

const DELETE_POINT = gql`mutation ($id: ID!){
  deletePoint(
    id: $id
  )
}`;

const EDIT_POINT = gql`mutation ($id: ID!, $title: String, $category: String, $type: String, $groupid: String, $lng: Float, $lat: Float){
  editPoint(
    id: $id
    title: $title
    category: $category
    type: $type
    groupid: $groupid
    lng: $lng
    lat: $lat
  )
}`;

const ADD_POINT = gql`mutation ($point: String!){
  addPoint(
    point: $point
  ) 
}`;
const ALL_POINTS = gql`query{allPoints}`;

const App = () => {
  const [map, setMap] = useState(null);
  const [tab, setTab] = useState(0);
  const [mobileOpen, setMobileOpen] = useState(false);
  const [features, setFeatures] = useState(null);
  const [selectedFeatures, setSelectedFeatures] = useState([]);
  const [navigation, setNavigation] = useState('App');
  const [snackbarOpen, setSnackbarOpen] = useState(false);
  const [snackbarTransition] = useState(() => function slide(props) { return <Slide {...props} direction="left" />; });
  const [snackbarMessage, setSnackbarMessage] = useState({ message: '', severity: 'info' });

  const classes = useStyles();

  const handleSnackbarMessage = ({ severity, message }) => {
    setSnackbarMessage({ severity, message });
    setSnackbarOpen(true);
  };

  const handleError = (error) => {
    handleSnackbarMessage({ severity: 'error', message: error.message });
    console.error(error);
  };

  const [deletePoint] = useMutation(DELETE_POINT, {
    onError: handleError,
    refetchQueries: [{ query: ALL_POINTS, notifyOnNetworkStatusChange: true }],
  });

  const [editPoint] = useMutation(EDIT_POINT, {
    onError: handleError,
    refetchQueries: [{ query: ALL_POINTS, notifyOnNetworkStatusChange: true }],
  });

  const [addPoint] = useMutation(ADD_POINT, {
    onError: handleError,
    refetchQueries: [{ query: ALL_POINTS, notifyOnNetworkStatusChange: true }],
  });

  const {
    loading: pointsLoading,
    error: pointsError,
    data: pointsData,
  } = useQuery(ALL_POINTS, { fetchPolicy: 'network-only' });

  useEffect(() => {
    const doIt = () => {
      if (map && pointsData) {
        const source = map.getSource('points');
        if (source) {
          const featuresData = JSON.parse(pointsData.allPoints).data;
          map.getSource('points').setData(featuresData);
          setFeatures(featuresData.features);
        }
      }
    };
    doIt();
  }, [map, pointsData]);

  const handleDrawerToggle = () => {
    setMobileOpen(!mobileOpen);
  };

  return (
    <ThemeProvider theme={Theme}>
      <div className={classes.root}>
        <CssBaseline />
        <Snackbar TransitionComponent={snackbarTransition} onClose={() => setSnackbarOpen(false)} anchorOrigin={{ vertical: 'top', horizontal: 'center' }} open={snackbarOpen} autoHideDuration={5000}>
          <Alert onClose={() => setSnackbarOpen(false)} severity={snackbarMessage.severity}>
            {snackbarMessage.message}
          </Alert>
        </Snackbar>
        <nav className={classes.drawer}>
          <Hidden smUp implementation="js">
            <Navigator
              PaperProps={{ style: { width: drawerWidth } }}
              variant="temporary"
              open={mobileOpen}
              onClose={handleDrawerToggle}
              navigation={navigation}
              setNavigation={setNavigation}
            />
          </Hidden>
          <Hidden xsDown implementation="css">
            <Navigator
              PaperProps={{ style: { width: drawerWidth } }}
              navigation={navigation}
              setNavigation={setNavigation}
            />
          </Hidden>
        </nav>
        <div className={classes.app}>
          <Header
            tab={tab}
            setTab={setTab}
            className={classes.header}
            onDrawerToggle={handleDrawerToggle}
          />
          <main className={classes.main}>
            <Map
              addPoint={addPoint}
              navigation={navigation}
              setMap={setMap}
              tab={tab}
              setTab={setTab}
              setFeatures={setFeatures}
              setSelectedFeatures={setSelectedFeatures}
            />
            <Points
              handleSnackbarMessage={handleSnackbarMessage}
              deletePoint={deletePoint}
              editPoint={editPoint}
              tab={tab}
              features={features}
              selectedFeatures={selectedFeatures}
              setSelectedFeatures={setSelectedFeatures}
            />
          </main>
          <footer className={classes.footer}>
            <Copyright />
          </footer>
        </div>
      </div>
    </ThemeProvider>
  );
};

export default App;
