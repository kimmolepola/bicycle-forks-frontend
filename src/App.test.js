import {
  render, screen, act, fireEvent,
} from '@testing-library/react';
import React from 'react';
import { MockedProvider } from '@apollo/client/testing';
import TestRenderer from 'react-test-renderer';

import App from './App';

jest.mock('./components/Map', () => () => <div />); // eslint-disable-line react/display-name
jest.mock('./components/MapMobile', () => () => <div />); // eslint-disable-line react/display-name

const mocks = [];

it('renders without error', async () => {
  let rendered;
  await act(async () => {
    rendered = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <App />
      </MockedProvider>,
    );
  });
  expect(rendered).toMatchSnapshot();
});
