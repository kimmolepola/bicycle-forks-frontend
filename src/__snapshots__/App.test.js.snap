// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without error 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="makeStyles-root-1"
        style="display: flex;"
      >
        <nav
          class="makeStyles-drawer-2"
        >
          <div
            class="MuiDrawer-root MuiDrawer-docked"
          >
            <div
              class="MuiPaper-root MuiDrawer-paper MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft MuiPaper-elevation0"
              style="width: 256px;"
            >
              <ul
                class="MuiList-root"
              >
                <li
                  class="MuiListItem-root makeStyles-firebase-11 makeStyles-item-9 makeStyles-itemCategory-10 MuiListItem-gutters"
                >
                  Safebike
                  <button
                    class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
                    style="position: absolute; right: 0px;"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </button>
                </li>
                <li
                  class="MuiListItem-root makeStyles-categoryHeader-7 MuiListItem-gutters"
                >
                  <div
                    class="MuiListItemText-root"
                  >
                    <span
                      class="MuiTypography-root MuiListItemText-primary makeStyles-categoryHeaderPrimary-8 MuiTypography-body1 MuiTypography-displayBlock"
                    >
                      Admin
                    </span>
                  </div>
                </li>
                <div
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiListItem-root makeStyles-item-9 makeStyles-itemActiveItem-12 MuiListItem-gutters MuiListItem-button"
                  role="button"
                  tabindex="0"
                >
                  <div
                    class="MuiListItemIcon-root makeStyles-itemIcon-14"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M15.5 5.5c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zM5 12c-2.8 0-5 2.2-5 5s2.2 5 5 5 5-2.2 5-5-2.2-5-5-5zm0 8.5c-1.9 0-3.5-1.6-3.5-3.5s1.6-3.5 3.5-3.5 3.5 1.6 3.5 3.5-1.6 3.5-3.5 3.5zm5.8-10l2.4-2.4.8.8c1.3 1.3 3 2.1 5.1 2.1V9c-1.5 0-2.7-.6-3.6-1.5l-1.9-1.9c-.5-.4-1-.6-1.6-.6s-1.1.2-1.4.6L7.8 8.4c-.4.4-.6.9-.6 1.4 0 .6.2 1.1.6 1.4L11 14v5h2v-6.2l-2.2-2.3zM19 12c-2.8 0-5 2.2-5 5s2.2 5 5 5 5-2.2 5-5-2.2-5-5-5zm0 8.5c-1.9 0-3.5-1.6-3.5-3.5s1.6-3.5 3.5-3.5 3.5 1.6 3.5 3.5-1.6 3.5-3.5 3.5z"
                      />
                    </svg>
                  </div>
                  <div
                    class="MuiListItemText-root"
                  >
                    <span
                      class="MuiTypography-root MuiListItemText-primary makeStyles-itemPrimary-13 MuiTypography-body1 MuiTypography-displayBlock"
                    >
                      App
                    </span>
                  </div>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </div>
                <hr
                  class="MuiDivider-root makeStyles-divider-15"
                />
              </ul>
            </div>
          </div>
        </nav>
        <div
          class="makeStyles-app-3"
        >
          <header
            class="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary MuiPaper-elevation0"
          >
            <div
              class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
            >
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-align-items-xs-center"
              >
                <div
                  class="MuiGrid-root MuiGrid-item"
                >
                  <button
                    aria-label="open drawer"
                    class="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-17 MuiIconButton-colorInherit"
                    style="display: none;"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                        />
                      </svg>
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </button>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
                />
              </div>
            </div>
          </header>
          <div
            class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary makeStyles-secondaryBar-16 MuiPaper-elevation0"
          >
            <div
              class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
            >
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-align-items-xs-center"
              >
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
                >
                  <h1
                    class="MuiTypography-root MuiTypography-h5 MuiTypography-colorInherit"
                  >
                    Bike parking areas
                  </h1>
                </div>
              </div>
            </div>
          </div>
          <div
            class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary makeStyles-secondaryBar-16 MuiPaper-elevation0"
          >
            <div
              class="MuiTabs-root"
            >
              <div
                class="MuiTabs-scroller MuiTabs-fixed"
                style="overflow: hidden;"
              >
                <div
                  class="MuiTabs-flexContainer"
                  role="tablist"
                >
                  <button
                    aria-selected="true"
                    class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit Mui-selected"
                    role="tab"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiTab-wrapper"
                    >
                      Map
                    </span>
                  </button>
                  <button
                    aria-selected="false"
                    class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
                    role="tab"
                    tabindex="-1"
                    type="button"
                  >
                    <span
                      class="MuiTab-wrapper"
                    >
                      Point
                    </span>
                  </button>
                </div>
                <span
                  class="PrivateTabIndicator-root-21 PrivateTabIndicator-colorSecondary-23 MuiTabs-indicator"
                  style="left: 0px; width: 0px;"
                />
              </div>
            </div>
          </div>
          <main
            class="makeStyles-main-4"
          >
            <div />
            <div
              style="display: none;"
            >
              Lorem ipsum
            </div>
          </main>
        </div>
      </div>
      <div
        style="display: none; flex: 1; position: fixed; width: 100%; height: 100%;"
      >
        <div />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="makeStyles-root-1"
      style="display: flex;"
    >
      <nav
        class="makeStyles-drawer-2"
      >
        <div
          class="MuiDrawer-root MuiDrawer-docked"
        >
          <div
            class="MuiPaper-root MuiDrawer-paper MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft MuiPaper-elevation0"
            style="width: 256px;"
          >
            <ul
              class="MuiList-root"
            >
              <li
                class="MuiListItem-root makeStyles-firebase-11 makeStyles-item-9 makeStyles-itemCategory-10 MuiListItem-gutters"
              >
                Safebike
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
                  style="position: absolute; right: 0px;"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </li>
              <li
                class="MuiListItem-root makeStyles-categoryHeader-7 MuiListItem-gutters"
              >
                <div
                  class="MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiListItemText-primary makeStyles-categoryHeaderPrimary-8 MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    Admin
                  </span>
                </div>
              </li>
              <div
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root makeStyles-item-9 makeStyles-itemActiveItem-12 MuiListItem-gutters MuiListItem-button"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemIcon-root makeStyles-itemIcon-14"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.5 5.5c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zM5 12c-2.8 0-5 2.2-5 5s2.2 5 5 5 5-2.2 5-5-2.2-5-5-5zm0 8.5c-1.9 0-3.5-1.6-3.5-3.5s1.6-3.5 3.5-3.5 3.5 1.6 3.5 3.5-1.6 3.5-3.5 3.5zm5.8-10l2.4-2.4.8.8c1.3 1.3 3 2.1 5.1 2.1V9c-1.5 0-2.7-.6-3.6-1.5l-1.9-1.9c-.5-.4-1-.6-1.6-.6s-1.1.2-1.4.6L7.8 8.4c-.4.4-.6.9-.6 1.4 0 .6.2 1.1.6 1.4L11 14v5h2v-6.2l-2.2-2.3zM19 12c-2.8 0-5 2.2-5 5s2.2 5 5 5 5-2.2 5-5-2.2-5-5-5zm0 8.5c-1.9 0-3.5-1.6-3.5-3.5s1.6-3.5 3.5-3.5 3.5 1.6 3.5 3.5-1.6 3.5-3.5 3.5z"
                    />
                  </svg>
                </div>
                <div
                  class="MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiListItemText-primary makeStyles-itemPrimary-13 MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    App
                  </span>
                </div>
                <span
                  class="MuiTouchRipple-root"
                />
              </div>
              <hr
                class="MuiDivider-root makeStyles-divider-15"
              />
            </ul>
          </div>
        </div>
      </nav>
      <div
        class="makeStyles-app-3"
      >
        <header
          class="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary MuiPaper-elevation0"
        >
          <div
            class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-align-items-xs-center"
            >
              <div
                class="MuiGrid-root MuiGrid-item"
              >
                <button
                  aria-label="open drawer"
                  class="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-17 MuiIconButton-colorInherit"
                  style="display: none;"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
              />
            </div>
          </div>
        </header>
        <div
          class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary makeStyles-secondaryBar-16 MuiPaper-elevation0"
        >
          <div
            class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-align-items-xs-center"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
              >
                <h1
                  class="MuiTypography-root MuiTypography-h5 MuiTypography-colorInherit"
                >
                  Bike parking areas
                </h1>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary makeStyles-secondaryBar-16 MuiPaper-elevation0"
        >
          <div
            class="MuiTabs-root"
          >
            <div
              class="MuiTabs-scroller MuiTabs-fixed"
              style="overflow: hidden;"
            >
              <div
                class="MuiTabs-flexContainer"
                role="tablist"
              >
                <button
                  aria-selected="true"
                  class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit Mui-selected"
                  role="tab"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiTab-wrapper"
                  >
                    Map
                  </span>
                </button>
                <button
                  aria-selected="false"
                  class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit"
                  role="tab"
                  tabindex="-1"
                  type="button"
                >
                  <span
                    class="MuiTab-wrapper"
                  >
                    Point
                  </span>
                </button>
              </div>
              <span
                class="PrivateTabIndicator-root-21 PrivateTabIndicator-colorSecondary-23 MuiTabs-indicator"
                style="left: 0px; width: 0px;"
              />
            </div>
          </div>
        </div>
        <main
          class="makeStyles-main-4"
        >
          <div />
          <div
            style="display: none;"
          >
            Lorem ipsum
          </div>
        </main>
      </div>
    </div>
    <div
      style="display: none; flex: 1; position: fixed; width: 100%; height: 100%;"
    >
      <div />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
